include/pybind11/cast.h

```diff
@@ -1009,21 +1009,76 @@ struct type_caster<T, enable_if_t<std::is_arithmetic<T>::value && !is_std_char_t
         return true;
     }
 
-    static handle cast(T src, return_value_policy /* policy */, handle /* parent */) {
-        if (std::is_floating_point<T>::value) {
-            return PyFloat_FromDouble((double) src);
-        } else if (sizeof(T) <= sizeof(ssize_t)) {
-            // This returns a long automatically if needed
-            if (std::is_signed<T>::value)
-                return PYBIND11_LONG_FROM_SIGNED(src);
-            else
-                return PYBIND11_LONG_FROM_UNSIGNED(src);
-        } else {
-            if (std::is_signed<T>::value)
-                return PyLong_FromLongLong((long long) src);
-            else
-                return PyLong_FromUnsignedLongLong((unsigned long long) src);
-        }
+    template <class A, class B>
+    struct is_explicitly_convertible
+    {
+        enum {value = std::is_same<A, B>::value || (std::is_constructible<B, A>::value && !std::is_convertible<A, B>::value)};
+    };
```

Comment:
This struct can be simplified considerably:
```C++
template <class A, class B> using is_explicitly_convertible = bool_constant<
    std::is_same<A, B>::value || (std::is_constructible<B, A>::value && !std::is_convertible<A, B>::value>;
```

Reply:
`std::bool_constant` is a relatively new addition in C++17 and would require a break. Nevertheless, it is a great suggestion and really valuable feedback. I went ahead and used `std::integral_constant<bool, ..>` instead in 416a43f.

Reply:
> But I'm curious why the not-implicitly-convertible check is needed here.

It makes sure the overload doesn't become ambiguous.